service/worker/scanner/build_ids/scavenger.go

```diff
@@ -0,0 +1,282 @@
+					PageSize:      input.TaskQueueListPageSize,
+					NextPageToken: heartbeat.TaskQueueNextPageToken,
+				})
+				if err != nil {
+					return err
+				}
+				for heartbeat.TaskQueueIdx < len(tqResponse.Entries) {
+					entry := tqResponse.Entries[heartbeat.TaskQueueIdx]
+					buildIdsToRemove, err := a.findBuildIdsToRemove(ctx, rateLimiter, heartbeat, ns, entry, input.BuildIdRemovalMinAge)
+					if err != nil {
+						return err
+					}
+					if len(buildIdsToRemove) > 0 {
+						_, err := a.matchingClient.UpdateWorkerBuildIdCompatibility(ctx, &matchingservice.UpdateWorkerBuildIdCompatibilityRequest{
+							NamespaceId: nsId,
+							TaskQueue:   entry.TaskQueue,
+							Operation: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_RemoveBuildIds_{
+								RemoveBuildIds: &matchingservice.UpdateWorkerBuildIdCompatibilityRequest_RemoveBuildIds{
+									KnownUserDataVersion: entry.UserData.Version,
+									BuildIds:             buildIdsToRemove,
+								},
+							},
+						})
+						if err != nil {
+							a.logger.Error("Failed to update task queue user data", tag.Error(err))
+							continue
+						}
+					}
+					heartbeat.TaskQueueIdx++
+					activity.RecordHeartbeat(ctx, heartbeat)
+				}
+				heartbeat.TaskQueueIdx = 0
+				heartbeat.TaskQueueNextPageToken = tqResponse.NextPageToken
+				if len(heartbeat.TaskQueueNextPageToken) == 0 {
+					break
+				}
+				activity.RecordHeartbeat(ctx, heartbeat)
+			}
+			heartbeat.NamespaceIdx++
+			activity.RecordHeartbeat(ctx, heartbeat)
+		}
+		heartbeat.NamespaceIdx = 0
+		heartbeat.NamespaceNextPageToken = nsResponse.NextPageToken
+		if len(heartbeat.NamespaceNextPageToken) == 0 {
+			break
+		}
+		activity.RecordHeartbeat(ctx, heartbeat)
+	}
+	return nil
+}
+
+// Process a single user data entry. Queries visibility for each build id and updates build id state with STATE_DELETED
+// (tombstone) for every build id that can safely be deleted.
+// Returns a list of build ids that were removed.
+func (a *Activities) findBuildIdsToRemove(
+	ctx context.Context,
+	rateLimiter quotas.RateLimiter,
+	heartbeat heartbeatDetails,
+	ns *namespace.Namespace,
+	entry *persistence.TaskQueueUserDataEntry,
+	removalMinAge time.Duration,
+) ([]string, error) {
+	versioningData := entry.UserData.Data.GetVersioningData()
+	var buildIdsToRemove []string
+	for setIdx, set := range versioningData.GetVersionSets() {
+		setActive := len(set.BuildIds)
+		for buildIdIdx, buildId := range set.BuildIds {
+			if buildId.State == persistencespb.STATE_DELETED {
+				setActive--
+				continue
+			}
+			// Set default
+			if buildIdIdx == len(set.BuildIds)-1 {
+				// Can't delete the queue default
+				if setIdx == len(versioningData.VersionSets)-1 {
+					continue
+				}
+				// There's another active build id in this set
+				if setActive > 1 {
+					continue
+				}
+			}
+			buildIdAge := time.Now().UnixMilli() - buildId.StateUpdateTimestamp.WallClock
+			if buildIdAge < removalMinAge.Milliseconds() {
```

Comment:
do we really need the min age thing anyway? if someone is adding a new default, it can't be deleted because it's the default. if they're adding to an older version, with no workflows then.. why are they doing that?